src/hardware/adlib_gold.cpp

```diff
@@ -0,0 +1,244 @@
+
+class StereoProcessor {
+public:
+	StereoProcessor(const uint16_t sample_rate);
+	~StereoProcessor();
+
+	void Reset() noexcept;
+	void ControlWrite(const TDA8425_Reg reg, const TDA8425_Register data) noexcept;
+	AudioFrame Process(const AudioFrame &frame) noexcept;
+
+	// prevent copying and assignment
+	StereoProcessor(const StereoProcessor &) = delete;
+	StereoProcessor &operator=(const StereoProcessor &) = delete;
+
+private:
+	TDA8425_Chip *chip = nullptr;
+};
+
+StereoProcessor::StereoProcessor(const uint16_t sample_rate) : chip(nullptr)
+{
+	chip = (TDA8425_Chip *)malloc(sizeof(TDA8425_Chip));
+
+	TDA8425_Chip_Ctor(chip);
+	TDA8425_Chip_Setup(chip,
+	                   sample_rate,
+	                   TDA8425_Pseudo_C1_Table[TDA8425_Pseudo_Preset_2],
+	                   TDA8425_Pseudo_C2_Table[TDA8425_Pseudo_Preset_2],
+	                   TDA8425_Tfilter_Mode_Disabled);
+	TDA8425_Chip_Reset(chip);
+	TDA8425_Chip_Start(chip);
+
+	Reset();
+}
+
+StereoProcessor::~StereoProcessor()
+{
+	TDA8425_Chip_Stop(chip);
+	TDA8425_Chip_Dtor(chip);
+}
+
+void StereoProcessor::Reset() noexcept
+{
+	constexpr auto volume_0db = 60;
+	constexpr auto bass_0db   = 6;
+	constexpr auto treble_0db = 6;
+
+	ControlWrite(TDA8425_Reg_VL, volume_0db);
+	ControlWrite(TDA8425_Reg_VR, volume_0db);
+	ControlWrite(TDA8425_Reg_BA, bass_0db);
+	ControlWrite(TDA8425_Reg_TR, treble_0db);
+
+	constexpr auto stereo_output = TDA8425_Selector_Stereo_1;
+	constexpr auto linear_stereo = TDA8425_Mode_LinearStereo
+	                            << TDA8425_Reg_SF_STL;
+
+	ControlWrite(TDA8425_Reg_SF, stereo_output & linear_stereo);
+}
+
+void StereoProcessor::ControlWrite(const TDA8425_Reg addr,
+                                   const TDA8425_Register data) noexcept
+{
+	TDA8425_Chip_Write(chip, static_cast<TDA8425_Address>(addr), data);
+}
+
+AudioFrame StereoProcessor::Process(const AudioFrame &frame) noexcept
+{
+	TDA8425_Chip_Process_Data data = {};
+
+	data.inputs[0][0] = frame.left;
+	data.inputs[1][0] = frame.left;
+	data.inputs[0][1] = frame.right;
+	data.inputs[1][1] = frame.right;
+
+	TDA8425_Chip_Process(chip, &data);
+
+	return {data.outputs[0], data.outputs[1]};
+}
+
+// AdLib Gold module
+// -----------------
+
+AdlibGold::AdlibGold(const uint16_t sample_rate)
+        : surround_processor(nullptr),
+          stereo_processor(nullptr)
+{
+	surround_processor = std::make_unique<SurroundProcessor>(sample_rate);
+	stereo_processor   = std::make_unique<StereoProcessor>(sample_rate);
+}
+
+AdlibGold::~AdlibGold() = default;
+
+void AdlibGold::StereoControlWrite(const TDA8425_Reg reg,
+                                   const TDA8425_Register data) noexcept
+{
+	stereo_processor->ControlWrite(reg, data);
+}
+
+void AdlibGold::SurroundControlWrite(const uint8_t val) noexcept
+{
+	surround_processor->ControlWrite(val);
+}
+
+void AdlibGold::Process(const int16_t *in, const uint32_t frames, float *out) noexcept
+{
+	auto frames_left = frames;
```

Comment:
maybe `_remaining` instead of `_left`, to disambiguate it from `.left`?

Reply:
Yes, that's better, changing.